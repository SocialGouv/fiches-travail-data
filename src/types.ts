// eslint-disable-next-line import/no-unresolved
import type { Node } from "unist";

export type ExtractedReference = {
  text: string;
  code: { name: string; id: string } | undefined;
};

export type Code = {
  name: string;
  id: string;
};

export type Reference = {
  text: string;
  fmt: string;
  id: string;
  code: Code | undefined;
};

export type ResolvedReferences = {
  [codeId: string]: {
    name: string;
    articles: { text: string; fmt: string; id: string }[];
  };
};

export type UnravelledReference = {
  text: string;
  fmt: string | undefined;
  code: Code;
};

export interface CodeArticle extends Node {
  data: {
    num: string;
    id: string;
  };
}
