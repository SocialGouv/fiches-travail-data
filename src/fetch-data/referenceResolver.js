/*
Here we resolve the references : 
Given an article (or a range) and its code (code du travail ou securite sociale), we search for its
actual id in the legi data corpus. 
*/

import { CODE_TRAVAIL, codesFullNames } from "@socialgouv/reference-article";
import find from "unist-util-find";
import visit from "unist-util-visit";

const codes = {};
Object.values(codesFullNames).forEach(({ id }) => {
  const code = require(`@socialgouv/legi-data/data/${id}.json`);
  codes[id] = code;
});

// duplicated in reference Extractor
const rangeMarkers = ["à", "à"];

const CODE_UNKNOWN = { id: "UNDEFINED" };
// shall we use "code du travail" by default ?
const DEFAULT_CODE = CODE_TRAVAIL;

// dumb convert article.data.num as integer for comparison
// each part up to MAX_DEPTH is padded with PAD_LENGTH
const PAD_LENGTH = 5; // left pad numbers to X chars
const MAX_DEPTH = 5; // max number of L432-1-1-1
// padding numbers : 2 -> "0002"
const leftPad = (num) => {
  let padded = "" + num;
  while (padded.length < PAD_LENGTH) {
    padded = "0" + padded;
  }
  return padded;
};
// transform articles into comparable integers
const asInt = (num) => {
  const parts = num
    .replace(/[^\d-]/g, "")
    .split("-")
    .map(leftPad);
  while (parts.length < MAX_DEPTH) {
    parts.push(leftPad(0));
  }
  const int = parseInt(parts.join(""));
  return int;
};

function getLegiDataRange(code, start, end) {
  // check if num is numerically after start. also check LRD prefix
  const isAfterStart = (node) =>
    asInt(node.data.num) >= asInt(start) &&
    node.data.num.charAt(0) === start.charAt(0);

  // check if num is numerically before end. also check LRD prefix
  const isBeforeEnd = (node) =>
    asInt(node.data.num) <= asInt(end) &&
    node.data.num.charAt(0) === end.charAt(0);

  const articles = [];
  visit(code, "article", (node) => {
    if (isAfterStart(node) && isBeforeEnd(node)) {
      articles.push(node);
    }
  });
  return articles;
}

function formatStartEnd(startRaw, endRaw) {
  // we need to identify special case where end ref is relative to start ref (e.g. L. 4733-9 à 11)
  // if there's nothing in common between end and start, we consider being in this special case

  const [startParts, endParts] = [startRaw, endRaw].map((a) =>
    a
      .replace(/\u2011/g, "-")
      .replace()
      .split("-")
      .map((p) => p.trim())
  );

  const letter = startParts[0].slice(0, 1);

  const startNums = Array.from(startParts);
  startNums[0] = startNums[0].replace(/\D/g, "");

  let endNums = Array.from(endParts);
  endNums[0] = endNums[0].replace(/\D/g, "");

  if (endNums.length == 1 && /^\d+$/.test(endParts[0])) {
    const endRange = endNums[0];
    endNums = Array.from(startNums.slice(0, -1));
    endNums.push(endRange);
  }

  return [letter + startNums.join("-"), letter + endNums.join("-")];
}

// in case of a range (like "L. 4733-9 à 4733-11"), we try to identify
// the articles implicitly included within the range
function unravelRange(range) {
  const mark = rangeMarkers.filter((a) => range.text.includes(a))[0];
  const rawParts = range.text.split(mark).map((p) => p.trim());

  const code = range.code ? range.code : DEFAULT_CODE;

  if (rawParts.length == 2 && code != CODE_UNKNOWN) {
    // objective is to identify starting and ending articles (with the legi data correct format)
    // then we can do a legi-data lookup
    const [startRaw, endRaw] = rawParts;
    const [startFMT, endFMT] = formatStartEnd(startRaw, endRaw);

    const unraveled = getLegiDataRange(codes[code.id], startFMT, endFMT).map(
      (a) => {
        const fmt = a.data.num;
        // keep original text for beginning and end
        let text;
        if (startFMT == fmt) {
          text = startRaw;
        } else if (endFMT == fmt) {
          text = endRaw;
        }
        return { ...(text && { text }), code, fmt };
      }
    );

    if (unraveled.length > 0) {
      return unraveled;
    }
  }

  // default in case of error, note that we explicitly set code to unknown
  // in order to identify range errors
  return range.text.split(mark).map((a) => {
    return { code: CODE_UNKNOWN, text: a.trim() };
  });
}

function formatArticle(article) {
  // remove dot and spaces + remove non digit trailing chars + replace unicode dash ‑ to standard -
  return article
    .toUpperCase()
    .replace(".", "")
    .replace(" ", "")
    .replace(/\D*$/, "")
    .replace(/\u2011/g, "-");
}

function resolveReference(ref) {
  let toResolve = [ref];
  if (rangeMarkers.filter((a) => ref.text.includes(a)).length != 0) {
    toResolve = unravelRange(ref);
  }

  return toResolve.map((a) => {
    // use default code if no defined
    const code =
      (a.code == CODE_UNKNOWN) | (a.code == undefined) ? DEFAULT_CODE : a.code;

    if (!a.fmt) a.fmt = formatArticle(a.text);

    if (code && code != CODE_UNKNOWN) {
      const article = find(
        codes[code.id],
        (node) => node.type === "article" && node.data.num === a.fmt
      );
      if (article) {
        a.cid = article.data.cid;
        a.id = article.data.id;
        a.code = code;
      } else {
        // not found in code
        a.code = CODE_UNKNOWN;
      }
    }
    return a;
  });
}

function resolveReferences(refs) {
  const resolvedRefs = refs.map((ref) => resolveReference(ref)).flat();

  const deduplicated = resolvedRefs.reduce((acc, art) => {
    // drop duplicated references
    const existing = acc
      .map((a) => [a.text, a.fmt])
      .flat()
      .filter((v) => v);

    if (!(existing.includes(art.fmt) || existing.includes(art.text))) {
      acc.push(art);
    }
    return acc;
  }, []);

  // group by code
  const grouped = deduplicated.reduce((acc, art) => {
    const { code, ...rawArticle } = art;
    const parsedCode = code ? code : CODE_UNKNOWN;

    if (!Object.keys(acc).includes(parsedCode.id)) {
      acc[parsedCode.id] = { articles: [], name: parsedCode.name };
    }

    acc[parsedCode.id].articles.push(rawArticle);

    return acc;
  }, {});
  return grouped;
}

export { resolveReferences };
