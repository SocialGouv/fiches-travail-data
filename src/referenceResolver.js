/*
Here we resolve the references : 
Given an article and its code (code du travail ou securite sociale), we search for its
actual id in the legi data corpus. 
*/

import find from "unist-util-find";
import { codesFullNames, CODE_TRAVAIL } from "./referenceExtractor";

const codes = {};
Object.values(codesFullNames).forEach(({ id }) => {
  const code = require(`@socialgouv/legi-data/data/${id}.json`);
  codes[id] = code;
});

// duplicated in reference Extractor
const rangeMarkers = ["à", "à"];

function unravelRange(range) {
  // TODO : deal with actual ranges (N to M)
  const mark = rangeMarkers.filter((a) => range.article.includes(a))[0];
  return range.article.split(mark).map((a) => {
    return { article: a.trim(), code: range.code };
  });
}

function formatArticle(article) {
  return article.replace(".", "").replace(" ", "");
}

function resolveReference(ref) {
  let toResolve = [ref];
  if (rangeMarkers.map((a) => ref.article.includes(a))) {
    toResolve = unravelRange(ref);
  }

  return toResolve.map((a) => {
    // by default we try to resolve code du travail
    const codeId = a.code ? a.code.id : CODE_TRAVAIL.id;
    const code = codes[codeId];
    if (code) {
      const formattedArticle = formatArticle(a.article);
      const article = find(
        code,
        (node) => node.type === "article" && node.data.num === formattedArticle
      );
      if (article) {
        a.id = article.data.id;
        a.fmt = formattedArticle;
        // case we guessed code du travail and found article
        if (!a.code) {
          a.code = CODE_TRAVAIL;
        }
      }
    }
    return a;
  });
}

function resolveReferences(refs) {
  return refs.map((ref) => resolveReference(ref)).flat();
}

export { resolveReferences };
