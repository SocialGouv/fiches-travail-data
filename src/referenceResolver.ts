/*
Here we resolve the references :
Given an article (or a range) and its code (code du travail ou securite sociale), we search for its
actual id in the legi data corpus.
*/

import find from "unist-util-find";
import visit from "unist-util-visit";
import { codesFullNames, CODE_TRAVAIL } from "./referenceExtractor";
import { asInt, rangeMarkers } from "./utils";
import type { Reference, ResolvedReferences } from "./types";
// FIXME(rmelisson) : not sure to understand why import Node is causing issue
// eslint-disable-next-line import/no-unresolved
import type { Node } from "unist";

const codes = Object.values(codesFullNames).reduce<{ [id: string]: Node }>(
  (memo, { id }) => {
    const code = require(`@socialgouv/legi-data/data/${id}.json`) as Node;
    return { [id]: code, ...memo };
  },
  {}
);

type CodeArticle = {
  data: {
    num: string;
    id: string;
  };
};

const CODE_UNKNOWN = { id: "UNDEFINED", name: "code undefined" };
// shall we use "code du travail" by default ?
const DEFAULT_CODE = CODE_TRAVAIL;

function getLegiDataRange(
  code: Node,
  start: string,
  end: string
): CodeArticle[] {
  // check if num is numerically after start. also check LRD prefix
  const isAfterStart = (article: CodeArticle): boolean =>
    asInt(article.data.num) >= asInt(start) &&
    article.data.num.charAt(0) === start.charAt(0);

  // check if num is numerically before end. also check LRD prefix
  const isBeforeEnd = (article: CodeArticle): boolean =>
    asInt(article.data.num) <= asInt(end) &&
    article.data.num.charAt(0) === end.charAt(0);

  const articles: CodeArticle[] = [];
  visit(code, "article", (node) => {
    // using untyped tree representation, only option is to cast here
    const article = (node as unknown) as CodeArticle;
    if (isAfterStart(article) && isBeforeEnd(article)) {
      articles.push(article);
    }
  });
  return articles;
}

function formatStartEnd(startRaw: string, endRaw: string): [string, string] {
  // we need to identify special case where end ref is relative to start ref (e.g. L. 4733-9 à 11)
  // if there's nothing in common between end and start, we consider being in this special case

  const [startParts, endParts] = [startRaw, endRaw].map((a) =>
    a
      .replace(/\u2011/g, "-")
      .split("-")
      .map((p) => p.trim())
  );

  const letter = startParts[0].slice(0, 1);

  const startNums = Array.from(startParts);
  startNums[0] = startNums[0].replace(/\D/g, "");

  let endNums = Array.from(endParts);
  endNums[0] = endNums[0].replace(/\D/g, "");

  if (endNums.length == 1 && /^\d+$/.test(endParts[0])) {
    const endRange = endNums[0];
    endNums = Array.from(startNums.slice(0, -1));
    endNums.push(endRange);
  }

  return [letter + startNums.join("-"), letter + endNums.join("-")];
}

type UnravelledReference = {
  text: string;
  fmt: string | undefined;
  code: {
    name: string;
    id: string;
  };
};

// in case of a range (like "L. 4733-9 à 4733-11"), we try to identify
// the articles implicitly included within the range
function unravelRange(range: Reference): UnravelledReference[] {
  const mark = rangeMarkers.filter((a) => range.text.includes(a))[0];
  const rawParts = range.text.split(mark).map((p) => p.trim());

  const code = range.code ? range.code : DEFAULT_CODE;

  if (rawParts.length == 2 && code != CODE_UNKNOWN) {
    // objective is to identify starting and ending articles (with the legi data correct format)
    // then we can do a legi-data lookup
    const [startRaw, endRaw] = rawParts;
    const [startFMT, endFMT] = formatStartEnd(startRaw, endRaw);

    const unraveled = getLegiDataRange(codes[code.id], startFMT, endFMT).map(
      (a) => {
        const fmt = a.data !== undefined ? a.data.num : undefined;
        // keep original text for beginning and end
        let text;
        if (startFMT == fmt) {
          text = startRaw;
        } else if (endFMT == fmt) {
          text = endRaw;
        } else {
          text = range.text + " => " + fmt;
        }
        return { text, fmt, code };
      }
    );

    if (unraveled.length > 0) {
      return unraveled;
    }
  }

  // default in case of error, note that we explicitly set code to unknown
  // in order to identify range errors
  return range.text.split(mark).map((a) => {
    return { text: a.trim(), fmt: undefined, code: CODE_UNKNOWN };
  });
}

function formatArticle(article: string): string {
  // remove dot and spaces + remove non digit trailing chars + replace unicode dash ‑ to standard -
  return article
    .replace(".", "")
    .replace(" ", "")
    .replace(/\D*$/, "")
    .replace(/\u2011/g, "-");
}

function resolveReference(ref: Reference): Reference[] {
  let toResolve: Reference[] = [ref];
  if (rangeMarkers.filter((a) => ref.text.includes(a)).length != 0) {
    toResolve = unravelRange(ref).map((ref) => ref as Reference);
  }

  return toResolve.map((a) => {
    // use default code if no defined
    const code =
      a.code == CODE_UNKNOWN || a.code == undefined ? DEFAULT_CODE : a.code;

    if (!a.fmt) a.fmt = formatArticle(a.text);

    if (code && code != CODE_UNKNOWN) {
      // again we have to cast as nodes are generic
      const article = (find(
        codes[code.id],
        (node: Node) => node.type === "article" && node.data?.num === a.fmt
      ) as unknown) as CodeArticle;
      if (article) {
        a.id = article.data ? article.data.id : "";
        a.code = code;
      } else {
        // not found in code
        a.code = CODE_UNKNOWN;
      }
    }
    return a;
  });
}

function resolveReferences(refs: Reference[]): ResolvedReferences {
  const resolvedRefs = refs.map(resolveReference).flat();

  const deduplicated = resolvedRefs.reduce((acc, art) => {
    // drop duplicated references
    const existing = acc
      .map((a: Reference) => [a.text, a.fmt])
      .flat()
      .filter((v) => v);

    if (!(existing.includes(art.fmt) || existing.includes(art.text))) {
      acc.push(art);
    }
    return acc;
  }, new Array<Reference>());

  // group by code
  const grouped = deduplicated.reduce((acc, art) => {
    const { code, ...rawArticle } = art;
    const parsedCode = code ? code : CODE_UNKNOWN;

    if (!acc.has(parsedCode.id)) {
      acc.set(parsedCode.id, { name: parsedCode.name, articles: [] });
    }
    acc.get(parsedCode.id)?.articles.push(rawArticle);

    return acc;
  }, new Map<string, { name: string; articles: { text: string; fmt: string; id: string }[] }>());
  return Object.fromEntries(grouped);
}

export { resolveReferences };
