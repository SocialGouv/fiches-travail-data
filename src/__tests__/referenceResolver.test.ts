import { resolveReferences } from "../referenceResolver";
import { extractReferences } from "../referenceExtractor";
import { Reference } from "../types";

test("should resolve les articles L. 2313‑8 et R. 2313-3 à R. 2313-6 du code du travail ainsi que le L. 1251-18", () => {
  expect(
    resolveReferences([
      {
        code: {
          id: "LEGITEXT000006072050",
          name: "code du travail",
        },
        text: "L. 2313‑8",
      },
      {
        code: {
          id: "LEGITEXT000006072050",
          name: "code du travail",
        },
        text: "R. 2313-3 à R. 2313-6",
      },
      {
        code: undefined,
        text: "L. 1251-18",
      },
    ] as Reference[])
  ).toMatchSnapshot();
});

test("should resolve L. 1251-23xx du code du travail", () => {
  expect(
    resolveReferences([
      {
        code: {
          id: "LEGITEXT000006072050",
          name: "code du travail",
        },
        text: "L. 1251-21 à L. 1251-23xx",
      },
    ] as Reference[])
  ).toMatchSnapshot();
});

test.each([
  "L. 1251-21 à L. 1251-23xx du code du travail",
  "L. 1233‑34 à L. 1233-35-1 du code du travail",
  "L. 2312-72 à 2312-77 du code du travail",
  "L. 2312-72 à 2312-77 du code de l'éducation",
  "D. 5132-9 à D. 5132-10-4 du code du travail",
  "D. 5132-9 à D. 5132-10-4",
  "L. 2315-38 à 40 du code du travail",
  "L. 351-1 à L. 351-5 du code de la sécurité sociale",
])('should resolve range "%s"', (range) => {
  const extractedRefs = extractReferences(range);
  const resolvedRefs = resolveReferences(extractedRefs);

  expect(resolvedRefs).toMatchSnapshot();
});
