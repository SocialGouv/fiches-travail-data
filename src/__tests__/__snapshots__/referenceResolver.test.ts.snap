// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should resolve L. 1251-23xx du code du travail 1`] = `
Object {
  "LEGITEXT000006072050": Object {
    "articles": Array [
      Object {
        "fmt": "L1251-21",
        "id": "LEGIARTI000006901274",
        "text": "L. 1251-21",
      },
      Object {
        "fmt": "L1251-22",
        "id": "LEGIARTI000006901275",
        "text": "L. 1251-21 à L. 1251-23xx => L1251-22",
      },
      Object {
        "fmt": "L1251-23",
        "id": "LEGIARTI000006901276",
        "text": "L. 1251-21 à L. 1251-23xx => L1251-23",
      },
    ],
    "name": "code du travail",
  },
}
`;

exports[`should resolve les articles L. 2313‑8 et R. 2313-3 à R. 2313-6 du code du travail ainsi que le L. 1251-18 1`] = `
Object {
  "LEGITEXT000006072050": Object {
    "articles": Array [
      Object {
        "fmt": "L2313-8",
        "id": "LEGIARTI000036262496",
        "text": "L. 2313‑8",
      },
      Object {
        "fmt": "R2313-3",
        "id": "LEGIARTI000039624940",
        "text": "R. 2313-3",
      },
      Object {
        "fmt": "R2313-4",
        "id": "LEGIARTI000036414723",
        "text": "R. 2313-3 à R. 2313-6 => R2313-4",
      },
      Object {
        "fmt": "R2313-5",
        "id": "LEGIARTI000039347715",
        "text": "R. 2313-3 à R. 2313-6 => R2313-5",
      },
      Object {
        "fmt": "R2313-6",
        "id": "LEGIARTI000039624937",
        "text": "R. 2313-6",
      },
      Object {
        "fmt": "L1251-18",
        "id": "LEGIARTI000006901269",
        "text": "L. 1251-18",
      },
    ],
    "name": "code du travail",
  },
}
`;

exports[`should resolve range "D. 5132-9 à D. 5132-10-4 du code du travail" 1`] = `
Object {
  "LEGITEXT000006072050": Object {
    "articles": Array [
      Object {
        "fmt": "D5132-10-1",
        "id": "LEGIARTI000029775717",
        "text": "D. 5132-9 à D. 5132-10-4 => D5132-10-1",
      },
      Object {
        "fmt": "D5132-10-2",
        "id": "LEGIARTI000029775713",
        "text": "D. 5132-9 à D. 5132-10-4 => D5132-10-2",
      },
      Object {
        "fmt": "D5132-10-3",
        "id": "LEGIARTI000029775708",
        "text": "D. 5132-9 à D. 5132-10-4 => D5132-10-3",
      },
      Object {
        "fmt": "D5132-10-4",
        "id": "LEGIARTI000029775703",
        "text": "D. 5132-10-4",
      },
    ],
    "name": "code du travail",
  },
}
`;

exports[`should resolve range "D. 5132-9 à D. 5132-10-4" 1`] = `
Object {
  "LEGITEXT000006072050": Object {
    "articles": Array [
      Object {
        "fmt": "D5132-10-1",
        "id": "LEGIARTI000029775717",
        "text": "D. 5132-9 à D. 5132-10-4 => D5132-10-1",
      },
      Object {
        "fmt": "D5132-10-2",
        "id": "LEGIARTI000029775713",
        "text": "D. 5132-9 à D. 5132-10-4 => D5132-10-2",
      },
      Object {
        "fmt": "D5132-10-3",
        "id": "LEGIARTI000029775708",
        "text": "D. 5132-9 à D. 5132-10-4 => D5132-10-3",
      },
      Object {
        "fmt": "D5132-10-4",
        "id": "LEGIARTI000029775703",
        "text": "D. 5132-10-4",
      },
    ],
    "name": "code du travail",
  },
}
`;

exports[`should resolve range "L. 351-1 à L. 351-5 du code de la sécurité sociale" 1`] = `
Object {
  "LEGITEXT000006073189": Object {
    "articles": Array [
      Object {
        "fmt": "L351-1",
        "id": "LEGIARTI000023035688",
        "text": "L. 351-1",
      },
      Object {
        "fmt": "L351-1-1",
        "id": "LEGIARTI000028498678",
        "text": "L. 351-1 à L. 351-5 => L351-1-1",
      },
      Object {
        "fmt": "L351-1-2",
        "id": "LEGIARTI000025014540",
        "text": "L. 351-1 à L. 351-5 => L351-1-2",
      },
      Object {
        "fmt": "L351-1-3",
        "id": "LEGIARTI000028498841",
        "text": "L. 351-1 à L. 351-5 => L351-1-3",
      },
      Object {
        "fmt": "L351-1-4",
        "id": "LEGIARTI000035640851",
        "text": "L. 351-1 à L. 351-5 => L351-1-4",
      },
      Object {
        "fmt": "L351-2",
        "id": "LEGIARTI000031686263",
        "text": "L. 351-1 à L. 351-5 => L351-2",
      },
      Object {
        "fmt": "L351-3",
        "id": "LEGIARTI000028498769",
        "text": "L. 351-1 à L. 351-5 => L351-3",
      },
      Object {
        "fmt": "L351-3-1",
        "id": "LEGIARTI000031667793",
        "text": "L. 351-1 à L. 351-5 => L351-3-1",
      },
      Object {
        "fmt": "L351-4",
        "id": "LEGIARTI000037063369",
        "text": "L. 351-1 à L. 351-5 => L351-4",
      },
      Object {
        "fmt": "L351-4-1",
        "id": "LEGIARTI000017845585",
        "text": "L. 351-1 à L. 351-5 => L351-4-1",
      },
      Object {
        "fmt": "L351-4-2",
        "id": "LEGIARTI000028497884",
        "text": "L. 351-1 à L. 351-5 => L351-4-2",
      },
      Object {
        "fmt": "L351-5",
        "id": "LEGIARTI000021537902",
        "text": "L. 351-5",
      },
    ],
    "name": "code de la sécurité sociale",
  },
}
`;

exports[`should resolve range "L. 1233‑34 à L. 1233-35-1 du code du travail" 1`] = `
Object {
  "LEGITEXT000006072050": Object {
    "articles": Array [
      Object {
        "fmt": "L1233-34",
        "id": "LEGIARTI000036762068",
        "text": "L. 1233‑34",
      },
      Object {
        "fmt": "L1233-35",
        "id": "LEGIARTI000036261799",
        "text": "L. 1233‑34 à L. 1233-35-1 => L1233-35",
      },
      Object {
        "fmt": "L1233-35-1",
        "id": "LEGIARTI000035643013",
        "text": "L. 1233-35-1",
      },
    ],
    "name": "code du travail",
  },
}
`;

exports[`should resolve range "L. 1251-21 à L. 1251-23xx du code du travail" 1`] = `
Object {
  "LEGITEXT000006072050": Object {
    "articles": Array [
      Object {
        "fmt": "L1251-21",
        "id": "LEGIARTI000006901274",
        "text": "L. 1251-21",
      },
      Object {
        "fmt": "L1251-22",
        "id": "LEGIARTI000006901275",
        "text": "L. 1251-21 à L. 1251-23xx => L1251-22",
      },
      Object {
        "fmt": "L1251-23",
        "id": "LEGIARTI000006901276",
        "text": "L. 1251-21 à L. 1251-23xx => L1251-23",
      },
    ],
    "name": "code du travail",
  },
}
`;

exports[`should resolve range "L. 2312-72 à 2312-77 du code de l'éducation" 1`] = `Object {}`;

exports[`should resolve range "L. 2312-72 à 2312-77 du code du travail" 1`] = `
Object {
  "LEGITEXT000006072050": Object {
    "articles": Array [
      Object {
        "fmt": "L2312-72",
        "id": "LEGIARTI000035611295",
        "text": "L. 2312-72",
      },
      Object {
        "fmt": "L2312-73",
        "id": "LEGIARTI000035611297",
        "text": "L. 2312-72 à 2312-77 => L2312-73",
      },
      Object {
        "fmt": "L2312-74",
        "id": "LEGIARTI000035611299",
        "text": "L. 2312-72 à 2312-77 => L2312-74",
      },
      Object {
        "fmt": "L2312-75",
        "id": "LEGIARTI000035611301",
        "text": "L. 2312-72 à 2312-77 => L2312-75",
      },
      Object {
        "fmt": "L2312-76",
        "id": "LEGIARTI000035611303",
        "text": "L. 2312-72 à 2312-77 => L2312-76",
      },
      Object {
        "fmt": "L2312-77",
        "id": "LEGIARTI000035611305",
        "text": "2312-77",
      },
    ],
    "name": "code du travail",
  },
}
`;

exports[`should resolve range "L. 2315-38 à 40 du code du travail" 1`] = `
Object {
  "LEGITEXT000006072050": Object {
    "articles": Array [
      Object {
        "fmt": "L2315-38",
        "id": "LEGIARTI000035626459",
        "text": "L. 2315-38",
      },
      Object {
        "fmt": "L2315-39",
        "id": "LEGIARTI000036262434",
        "text": "L. 2315-38 à 40 => L2315-39",
      },
      Object {
        "fmt": "L2315-40",
        "id": "LEGIARTI000035626463",
        "text": "40",
      },
    ],
    "name": "code du travail",
  },
}
`;
